pkg/base/helm.go

```diff
@@ -243,57 +254,168 @@ func removeCommonPrefix(baseFiles []BaseFile) []BaseFile {
+	}
 
 	var nextBases []Base
-	for _, base := range base.Bases {
-		base = helmChartBaseAppendAdditionalFiles(base, u)
-		nextBases = append(nextBases, base)
+	for _, nextBase := range base.Bases {
+		nextBase = helmChartBaseAppendAdditionalFiles(nextBase, path.Join(base.Path, nextBase.Path), upstreamFiles, baseToUpstreamPathMap)
+		nextBases = append(nextBases, nextBase)
 	}
 	base.Bases = nextBases
 
 	return base
 }
 
-// look for any sub-chart dependencies that are missing from base and add their Chart.yaml
-func helmChartBaseAppendMissingDependencies(base Base, u *upstreamtypes.Upstream) Base {
-	basePaths := getAllBasePaths("", base)
-	basePathMap := map[string]bool{}
-	for _, basePath := range basePaths {
-		basePathMap[basePath] = true
+func helmChartUpstreamPathToBasePaths(upstreamPath string, upstreamFileMap map[string][]byte) ([]string, error) {
+	charts := pathToCharts(upstreamPath)
+	basePathParts := make([][]string, len(charts))
+	basePathParts[0] = []string{""}
+	// iterate over the subcharts and find any that are aliased
+	for i := 1; i < len(charts); i++ {
+		parts := []string{}
+		chart := charts[i]
+		parts = append(parts, chart)
+		parentChart := charts[i-1]
+
+		// build the full parent chart path and check if the chart is aliased by its parent
+		parentChartYaml := filepath.Join(parentChart, "Chart.yaml")
+		for j := 0; j < i-1; j++ {
+			parentChartYaml = filepath.Join("charts", parentChartYaml)
+		}
+		if content, ok := upstreamFileMap[parentChartYaml]; ok {
+			deps := new(HelmChartDependencies)
+			if err := yaml.Unmarshal(content, deps); err != nil {
+				return nil, errors.Wrapf(err, "failed to unmarshal %s", parentChartYaml)
+			}
+			for _, dep := range deps.Dependencies {
+				// replace any aliased subcharts with the actual chart name
+				if dep.Name == chart && dep.Alias != "" {
+					parts = append(parts, dep.Alias)
+				}
+			}
+		} else {
+			return nil, errors.Errorf("failed to find upstream file %s", parentChartYaml)
+		}
+
+		basePathParts[i] = parts
 	}
 
-	for _, upstreamFile := range u.Files {
-		basePath := strings.TrimSuffix(upstreamFile.Path, "Chart.yaml")
-		basePath = strings.TrimSuffix(basePath, "/")
-		if !basePathMap[basePath] && strings.HasSuffix(upstreamFile.Path, "Chart.yaml") {
+	return flattenBasePathParts(basePathParts), nil
+}
+
+// Takes an input chart path and returns a list that represents the dependency tree for the chart.
+// The top-level chart is represented by an empty string.
+//  // "" => [""]
+//  // "charts/mariadb" => ["", "mariadb"]
+//  // "charts/mariadb/charts/common" => ["", "mariadb", "common"]
+func pathToCharts(path string) []string {
+	re := regexp.MustCompile(`\/?charts\/`)
+	return re.Split(path, -1)
+}
+
+// takes a slice of string slices representing the aliased base paths for each level of a chart
+// and returns a slice of strings representing all of the possible base paths
+func flattenBasePathParts(basePathParts [][]string) []string {
```

Comment:
if i understand the purpose of this function correctly, each index of `basePathParts` contains a list of aliases/names for the subcharts. wouldn't it be simpler to flatten those with a recursive function? i don't quite grasp the implementation of this to be honest.

Reply:
same as above